sympy/matrices/matrices.py

```diff
@@ -174,10 +175,24 @@ def _eval_det_bareiss(self):
         """
 
         # XXX included as a workaround for issue #12362.  Should use `_find_reasonable_pivot` instead
+        # As of 2018-01-09, `_find_reasonable_pivot` could be used but it slows down the computation,
+        # by the factor of 2.5 in one test. The issue #10279 is relevant.
+        # As a compromise, check for the value being nonzero at several random rational points
         def _find_pivot(l):
-            for pos,val in enumerate(l):
+            for pos, val in enumerate(l):
                 if val:
-                    return (pos, val, None, None)
+                    variables = val.free_symbols
+                    good_pivot = True
+                    if len(variables) > 0:
+                        for _ in range(5):
+                            substitutions = {v: random.randint(-10, 10) + Rational(
+                                random.randint(100, 200), random.randint(201, 300)) \
```

Comment:
see also `Expr._random`. An alternative would be to give `zero` a `quick` option so it would exit before it starts any heavy lifting.

Reply:
`Expr._random()` uses floats and tries to compute the expression with precision, which can lead to serious performance issues with complex expressions where a lot of cancellation takes place. This is an example extracted from the determinant that in the linked issue:
```
var('a')
expr = ((-4*a*(-4*a*(a + 1) + a*(4*a + 1)) + a*(-4*a*(a + 4) + a*(4*a + 4)))*(-2*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 2) + a*(2*a + 2)))/a**2 - (-2*a*(-4*a*(a + 1) + a*(4*a + 1)) + a*(-4*a*(a + 2) + a*(4*a + 2)))*(-4*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 4) + a*(2*a + 4)))/a**2)*((-3*a*(-3*a*(a + 1) + a*(3*a + 1)) + a*(-3*a*(a + 3) + a*(3*a + 3)))*(-2*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 2) + a*(2*a + 2)))/a**2 - (-2*a*(-3*a*(a + 1) + a*(3*a + 1)) + a*(-3*a*(a + 2) + a*(3*a + 2)))*(-3*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 3) + a*(2*a + 3)))/a**2)/a**2 - ((-3*a*(-4*a*(a + 1) + a*(4*a + 1)) + a*(-4*a*(a + 3) + a*(4*a + 3)))*(-2*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 2) + a*(2*a + 2)))/a**2 - (-2*a*(-4*a*(a + 1) + a*(4*a + 1)) + a*(-4*a*(a + 2) + a*(4*a + 2)))*(-3*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 3) + a*(2*a + 3)))/a**2)*((-4*a*(-3*a*(a + 1) + a*(3*a + 1)) + a*(-3*a*(a + 4) + a*(3*a + 4)))*(-2*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 2) + a*(2*a + 2)))/a**2 - (-2*a*(-3*a*(a + 1) + a*(3*a + 1)) + a*(-3*a*(a + 2) + a*(3*a + 2)))*(-4*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 4) + a*(2*a + 4)))/a**2)/a**2
```
Here, `expr.subs(a, Rational(123, 234))`  (or another rational number) returns 0 instantly. I timed `expr._random()` at 122 seconds on my laptop; it was a good workout for the CPU. (The returned value as None.)

I'm not sure what `zero` method you refer to, what module is it in? I took a look at `equals` method of Expr (which at some point is using `_random`, but only after simplification). Perhaps this is where quick option could go?  

There is also the matter of how much one is worried about erroneous classification. In case of a pivot, it's really bad to pick zero element as a pivot, but rejecting a nonzero value is no big deal; there are probably other entries that can be used instead.  For other purposes, the relative importance of two kinds of failure may be different.